lib/fileset/README.md

```diff
@@ -0,0 +1,165 @@
+
+### Influence tracking
+
+File set operations internally track the top-most directory that could influence the included files.
+And `toSource` requires `fileset` to not be influencable by files outside of `root`.
+For example, even with `dir/file.txt` being the only file in `./.`, `toSource { root = ./dir; fileset = ./.; }` gives an error.
+This is because the file set could be influenced by any files in `./.`, which is outside `./dir`.
+
+Arguments:
+- (+) This gives us the guarantee that adding new files to a project never breaks a file set expression.
+  This also exists in a lesser form for removed files, where only removing files referenced by paths coerced to file sets can break a file set expression.
+- (+) This can be removed later, if we discover it's too restrictive
+- (-) It leads to errors when a sensible result could be returned, such as in the above example.
+
+### Empty directories
+
+File sets strictly can only represent a _set_ of local files, directories on their own are not representable.
+
+Arguments:
+- (+) There does not seem to be a sensible set of combinators when directories can be represented on their own.
+  Here's some possibilities:
+  - `./.` represents the files in `./.` _and_ the directory itself including its subdirectories, meaning that even if there's no files, the entire structure of `./.` is preserved
+
+    In that case, what should `fileFilter (file: false) ./.` return?
+    It could return the entire directory structure unchanged, but with all files removed, which would not be what one would expect.
+
+    Trying to have a filter function that also supports directories will lead to the question of:
+    What should the behavior be if `./foo` itself is excluded but all of its contents are included?
+    It leads to having to define when directories are recursed into, but then we're effectively back at how the `builtins.path`-based filters work.
+
+  - `./.` represents all files in `./.` _and_ the directory itself, but not its subdirectories, meaning that at least `./.` will be preserved even if it's empty.
+
+    In that case, `intersect ./. ./foo` should only include files and no directories themselves, since `./.` includes only `./.` as a directory, and same for `./foo`, so there's no overlap in directories.
+    But intuitively this operation should result in the same as `./foo`, that not being the case is just confusing.
+- (+) This matches Git also only supporting files, so developers are already used to it
+- (-) Empty directories (even if they contain nested directories) are neither representable nor preserved when coercing from paths.
+  - (+) It is very rare that empty directories are necessary.
+  - (+) We can implement a workaround, allowing `toSource` to take an extra argument for ensuring certain extra directories exist in the result.
+- (-) It slows down store imports, since the evaluator needs to traverse the entire tree to remove any empty directories
```

Comment:
Having done some brief benchmarks, it turns out this might actually be very significant. While it does make the design nice to strictly never have empty directories, in practice almost nobody uses or cares about empty directories. So we'd be paying a significant performance cost for an insignificant benefit. It could be optimized in Nix itself, but even that takes considerable effort with the same insignificant benefit.

The one decent alternative I can think of is to not remove empty directories within directories that don't have to be recursed into. So e.g. `./.` as a fileset would not remove any empty directories. but `fileFilter pred ./.` would, because `fileFilter` has to recurse into the entire thing.

I will add a proper benchmark for this so we can reason about this easier.

Reply:
Thinking about this some more, I don't think that alternative is okay. It leads to unexpected and hard-to-explain semantics which we could not get away from. The right solution is to have better support for this in Nix itself, e.g. by adjusting `builtins.path` as I just suggested here: https://github.com/NixOS/nix/issues/8820

Reply:
Note to self: Reuse the benchmark and Nix code from https://github.com/NixOS/nix/issues/8820#issuecomment-1678256565, maybe even export the Nix code as a new function under `lib.sources`, since it's useful on its own and could use a bunch of testing, the implementation is very non-trivial. It would also remove the nastiest code from the file set combinators implementation.

Reply:
Alright this should be fine, as discussed in the above issue, it's I/O bottlenecked, and while there is increased memory usage, it's not _too_ bad. I also added a benchmark now, allowing comparing to previous versions.